### DISCUSSION TITLE
Using groq as a chat completion service in semantic kernel

### DISCUSSION THREAD

**Discussion by giannik** [microsoft/semantic-kernel#9623] [Q&A] [answered]
What would be the recommended way in semantic kernel to consume  groq api for chatCompletion ?
I found this article suggesting a httpClient with a custom delegate handler:
[HttpClient with a custom delegate handler](https://www.linkedin.com/pulse/using-groq-from-semantic-kernel-andr%C3%A1s-velv%C3%A1rt-suhnf/)

but i also discovered  Microsoft.Extensions.AI that recommends using an IChatClient :
https://devblogs.microsoft.com/dotnet/introducing-microsoft-extensions-ai-preview/

What would be the recommended way in SK. @RogerBarreto

**Comment by RogerBarret0** [comment, marked as answer]
@giannik Keep in mind that we don't have an official connector for groq, and normally for API's that follows the OpenAI Chat Completion Standard, for simple tasks using the OpenAI connector may work as expected.

> [!Important]
> Semantic Kernel is very extensible and there's not a definitive answer, you can use below approaches

## OpenAI Custom Endpoint

In the first example you don't necessarily need a `httphandler`, as the OpenAI SDK allows custom endpoints to be passed.

We also have some UnitTest's ensuring this, see:

https://github.com/microsoft/semantic-kernel/blob/ffac88a83063c8b8444b7f712895dfae90b6b08e/dotnet/src/Connectors/Connectors.OpenAI.UnitTests/Services/OpenAIChatCompletionServiceTests.cs#L87

Using a custom endpoint:

```csharp
var chatCompletion = new OpenAIChatCompletionService(
    modelId: "any", 
    apiKey: "api-key", 
    endpoint: new Uri("https://api.groq.com/openai/v1"))
```

## AI Abstractions 
Using the example provided in the blog you can convert the `IChatClient` to `IChatCompletionService` easily like below:

```csharp
using Microsoft.SemanticKernel;

// <ProviderChatClient> here can be any, OllamaChatClient, OpenAIChatClient, etc... for groq, OpenAI may work best.
var service =  new <ProviderChatClient>(...)
    .AsChatCompletionService();


// your SK code
```

## Http Handler

Follow the approach mentioned in the link [HttpClient with a custom delegate handler](https://www.linkedin.com/pulse/using-groq-from-semantic-kernel-andr%C3%A1s-velv%C3%A1rt-suhnf/)

**Reply by giannik** [reply]
@RogerBarret0 thanks

**Comment by chorinator** [comment]
Following the IChatClient approach, I wrote an adapter by taking advantage of Groq's compatibility with OpenAI's API.
I used this [YouTube](https://youtu.be/30JTMFdsuKI?si=l3Bjx5oZOolzccfi) video as a reference.
Maybe you can follow a similar path for a SemanticKernel approach.
Hope it helps.

### Code
```csharp
using System.ClientModel;
using Microsoft.Extensions.AI;
using Microsoft.Extensions.Options;
using OpenAI;

namespace Adapters;

public class GroqChatClientAdapter : IChatClient
{
    private readonly IChatClient _client;

    public GroqChatClientAdapter(IOptions<GroqConfiguration> configuration)
    {
        var groqConfiguration = configuration.Value;
        var options = new OpenAIClientOptions
        {
            Endpoint = new Uri(groqConfiguration.Endpoint)
        };

        var client =
            new OpenAIClient(
                    new ApiKeyCredential(groqConfiguration.ApiKey),
                    options)
                .AsChatClient(groqConfiguration.Model);

        _client = client;
    }

    public void Dispose()
    {
        _client.Dispose();
    }

    public Task<ChatResponse> GetResponseAsync(IList<ChatMessage> chatMessages, ChatOptions? options = null,
        CancellationToken cancellationToken = new()) =>
        _client.GetResponseAsync(chatMessages, options, cancellationToken);

    public IAsyncEnumerable<ChatResponseUpdate> GetStreamingResponseAsync(IList<ChatMessage> chatMessages,
        ChatOptions? options = null,
        CancellationToken cancellationToken = new()) =>
        _client.GetStreamingResponseAsync(chatMessages, options, cancellationToken);

    public object? GetService(Type serviceType, object? serviceKey = null) =>
        _client.GetService(serviceType, serviceKey);
}
```

### Dependency Injection
```csharp
    public static void AddGroqChatClient(this IServiceCollection services, IConfiguration configuration)
    {
        services.AddChatClient(c =>
            ActivatorUtilities.CreateInstance<GroqChatClientAdapter>(c));
        services.Configure<GroqConfiguration>(configuration.GetSection("GroqConfiguration"));
    }
```

### Configuration class
```csharp
public class GroqConfiguration
{
    public string Endpoint { get; init; } = string.Empty;
    public string ApiKey { get; init; } = string.Empty;
    public string Model { get; init; } = string.Empty;
}
```
